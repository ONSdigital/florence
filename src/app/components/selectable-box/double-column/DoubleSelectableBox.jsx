import React from "react";
import PropTypes from "prop-types";
import clsx from "clsx";
import { useSort } from "../../../hooks/useSort";
import Sort from "../../sort";
import DoubleSelectableBoxItem from "./DoubleSelectableBoxItem";

const DoubleSelectableBox = props => {
    const { sortedItems, requestSort, sortConfig } = useSort(props.items, { key: "name", direction: "ASC" });
    const renderList = () => (
        <ul id="selectable-box" className="selectable-box__list" data-testid="selectable-box">
            {sortedItems.map(item => {
                return (
                    <DoubleSelectableBoxItem
                        key={item.id}
                        {...item}
                        isSelected={props.activeItemID && item.id === props.activeItemID}
                        handleClick={props.handleItemClick}
                    />
                );
            })}
        </ul>
    );
    const renderMessage = () => {
        if (props.search) {
            return (
                <div className="selectable-box__message">
                    <p>
                        <strong>Cannot find collection</strong>
                    </p>
                    <p>Improve your results by:</p>
                    <ul>
                        <li>double-checking your spelling</li>
                        <li>searching for something less specific</li>
                    </ul>
                </div>
            );
        } else {
            return (
                <div className="selectable-box__message">
                    <p>No items to display</p>
                </div>
            );
        }
    };

    return (
        <div className="selectable-box">
            <div className="grid">
                <button
                    aria-label="Sort by name"
                    className="selectable-box__heading with-sort grid__col-6 padding-right--0 grid__cell"
                    onClick={() => requestSort("name")}
                >
                    {props.headings[0]}
                    <Sort active={sortConfig} name="name" />
                </button>
                <button
                    aria-label="Sort by publishDate"
                    className="selectable-box__heading with-sort grid__col-6 grid__cell"
                    onClick={() => requestSort("publishDate")}
                >
                    {props.headings[1]}
                    <Sort active={sortConfig} name="publishDate" />
                    {props.isUpdating && <span className="selectable-box__status pull-right loader" />}
                </button>
            </div>
            {props.items.length > 0 ? renderList() : renderMessage()}
        </div>
    );
};

DoubleSelectableBox.propTypes = {
    headings: PropTypes.array.isRequired,
    search: PropTypes.string,
    items: PropTypes.arrayOf(
        PropTypes.shape({
            id: PropTypes.string.isRequired,
            selectableBox: PropTypes.shape({
                firstColumn: PropTypes.string.isRequired,
                secondColumn: PropTypes.string.isRequired,
            }),
        })
    ).isRequired,
    activeItemID: PropTypes.string,
    handleItemClick: PropTypes.func.isRequired,
    isUpdating: PropTypes.bool,
};

export default DoubleSelectableBox;
