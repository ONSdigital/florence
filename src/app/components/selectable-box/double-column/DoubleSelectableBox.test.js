import React from "react";
import { render, fireEvent, screen, getByText } from "@testing-library/react";
import { within } from "@testing-library/dom";
import "@testing-library/jest-dom/extend-expect";
import renderer from "react-test-renderer";
import { items } from "../../../../tests/mockData";
import DoubleSelectableBox from "./DoubleSelectableBox";

const props = {
    headings: ["Header1", "Header2"],
    items,
    activeItemID: items[1].id,
    isUpdating: false,
    search: "baz",
    handleItemClick: jest.fn(),
};

describe("DoubleSelectableBox", () => {
    it("matches the snapshot", () => {
        const component = renderer.create(<DoubleSelectableBox {...props} />);
        expect(component.toJSON()).toMatchSnapshot();
    });

    describe("when there are no items", () => {
        test("displays search result message not generic inside the table if search phrase was passed", () => {
            const emptyProps = {
                ...props,
                items: [],
            };

            const { container } = render(<DoubleSelectableBox {...emptyProps} />);
            screen.getByText("Header1");
            screen.getByText("Header2");

            expect(container).not.toHaveTextContent("No items to display");

            screen.getByText("Cannot find collection");
            screen.getByText("Improve your results by:");
            screen.getByText("double-checking your spelling");
            screen.getByText("searching for something less specific");
        });

        test("displays generic message when no search phrase was passed", () => {
            const emptyProps = {
                ...props,
                search: "",
                items: [],
            };
            render(<DoubleSelectableBox {...emptyProps} />);

            screen.getByText("No items to display");
            expect(screen.findByText("Cannot find collection")).not.toBeInTheDocument;
            expect(screen.findByText("Bar")).not.toBeInTheDocument;
        });
    });

    describe("when there are items passed in props", () => {
        test("displays items in the table", () => {
            const { container } = render(<DoubleSelectableBox {...props} />);
            const items = container.getElementsByClassName("selectable-box__item");

            expect(items.length).toBe(3);
            expect(items[0]).toHaveTextContent("Bar");
            expect(items[0]).toHaveTextContent("Friday");
            expect(items[1]).toHaveTextContent("Baz");
            expect(items[1]).toHaveTextContent("[manual collection]");
            expect(items[2]).toHaveTextContent("Foo");
            expect(items[2]).toHaveTextContent("2021-12-17T09:30:00.000Z");

            expect(container).not.toHaveTextContent("No items to display");
            expect(container).not.toHaveTextContent("Cannot find collection");
        });
    });

    describe("when Name header is clicked", () => {
        test("updates active sort to DESC and back to ASC", () => {
            render(<DoubleSelectableBox {...props} />);

            const sortByNameBtn = screen.getByRole("button", { name: /sort by name/i });
            const ascending = within(sortByNameBtn).getByTestId("ASC");
            const descending = within(sortByNameBtn).getByTestId("DESC");

            // sort by name is active by default
            expect(ascending).toHaveClass("active");
            expect(descending).not.toHaveClass("active");

            fireEvent.click(sortByNameBtn);

            expect(ascending).not.toHaveClass("active");
            expect(descending).toHaveClass("active");

            fireEvent.click(sortByNameBtn);

            expect(ascending).toHaveClass("active");
            expect(descending).not.toHaveClass("active");
        });
    });

    describe("when Publish Date header is clicked", () => {
        test("updates active sort to ASC and back to DESC by publish date column", () => {
            render(<DoubleSelectableBox {...props} />);

            const sortByPublishNameBtn = screen.getByRole("button", { name: /sort by publishdate/i });
            const ascending = within(sortByPublishNameBtn).getByTestId("ASC");
            const descending = within(sortByPublishNameBtn).getByTestId("DESC");

            // sort by publish date is not active by default
            expect(ascending).not.toHaveClass("active");
            expect(descending).not.toHaveClass("active");

            fireEvent.click(sortByPublishNameBtn);

            expect(ascending).not.toHaveClass("active");
            expect(descending).toHaveClass("active");

            fireEvent.click(sortByPublishNameBtn);

            expect(descending).not.toHaveClass("active");
            expect(ascending).toHaveClass("active");
        });
    });
});
