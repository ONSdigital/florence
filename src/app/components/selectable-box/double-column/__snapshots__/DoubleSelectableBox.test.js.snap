// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DoubleSelectableBox matches the snapshot 1`] = `
<div
  className="selectable-box"
>
  <div
    className="grid"
  >
    <button
      aria-label="Sort by name"
      className="selectable-box__heading with-sort grid__col-6 padding-right--0 grid__cell"
      onClick={[Function]}
    >
      Header1
      <span
        className="sort"
      >
        <span
          className="sort__icon-up active"
          data-testid="ASC"
        />
        <span
          className="sort__icon-down"
          data-testid="DESC"
        />
      </span>
    </button>
    <button
      aria-label="Sort by publishDate"
      className="selectable-box__heading with-sort grid__col-6 grid__cell"
      onClick={[Function]}
    >
      Header2
      <span
        className="sort"
      >
        <span
          className="sort__icon-up"
          data-testid="ASC"
        />
        <span
          className="sort__icon-down"
          data-testid="DESC"
        />
      </span>
    </button>
  </div>
  <ul
    className="selectable-box__list"
    data-testid="selectable-box"
    id="selectable-box"
  >
    <li
      className="selectable-box__item selected neutral"
      data-testid="2"
      id="2"
      onClick={[Function]}
    >
      <div
        className="grid"
      >
        <div
          className="grid__col-6"
        >
          Bar
          
        </div>
        <div
          className="grid__col-6"
        >
          Friday
        </div>
      </div>
    </li>
    <li
      className="selectable-box__item neutral"
      data-testid="3"
      id="3"
      onClick={[Function]}
    >
      <div
        className="grid"
      >
        <div
          className="grid__col-6"
        >
          Baz
          
        </div>
        <div
          className="grid__col-6"
        >
          [manual collection]
        </div>
      </div>
    </li>
    <li
      className="selectable-box__item neutral"
      data-testid="1"
      id="1"
      onClick={[Function]}
    >
      <div
        className="grid"
      >
        <div
          className="grid__col-6"
        >
          Foo
          
        </div>
        <div
          className="grid__col-6"
        >
          2021-12-17T09:30:00.000Z
        </div>
      </div>
    </li>
  </ul>
</div>
`;
